### PR TITLE
[mono] Implement Sse2.AddSaturate using LLVM

### PR DESCRIPTION
**Pull request by EgorBo**
Mono currently supports a limited subset of **Sse1**-**Sse42** intrinsics used only by CoreLib internally. So once a new API is used to optimize things there, we also have to implement it in Mono. So it recently happened in https://github.com/dotnet/runtime/pull/32036 (`Sse2.AddSaturate` was used for the first time).

So this PR implements it for mono (with all Sse2 overloads) using named LLVM intrinsics. It turns out it's different between LLVM6 (we currently based on) and LLVM9 (we plan to migrate to soon).

for `Vector128<byte>` overload it emits
```llvm
%result = call <8 x i16> @llvm.x86.sse2.paddus.b(<16 x i8> %left, <16 x i8> %right)
```
which is then emitted as `vpaddusw`

### CODE REVIEW

File: src/mono/mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -9359,6 +9385,18 @@ static IntrinsicDesc intrinsics[] = {
 	{INTRINS_SSE_ROUNDPD, "llvm.x86.sse41.round.pd"},
 	{INTRINS_SSE_PTESTZ, "llvm.x86.sse41.ptestz"},
 	{INTRINS_SSE_INSERTPS, "llvm.x86.sse41.insertps"},
+#if LLVM_API_VERSION >= 800
```
Comment (EgorBo) [inline src/mono/mono/mini/mini-llvm.c]: See https://godbolt.org/z/zNovw7

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/mini-ops.h [hunk 2]
```diff
@@ -1021,6 +1021,7 @@ MINI_OP(OP_SSE_AND, "sse_and", XREG, XREG, XREG)
 MINI_OP(OP_SSE_OR, "sse_or", XREG, XREG, XREG)
 MINI_OP(OP_SSE_XOR, "sse_xor", XREG, XREG, XREG)
 MINI_OP(OP_SSE_ANDN, "sse_andn", XREG, XREG, XREG)
+MINI_OP(OP_SSE_ADDS, "sse_adds", XREG, XREG, XREG)
```
Comment (imhameed) [inline src/mono/mono/mini/mini-ops.h]: Minor nitpick: paddsw/paddsb are SSE2, not SSE.

**Review by imhameed — APPROVED** [review]

File: src/mono/mono/mini/mini-ops.h (hunk 2 above)
Comment (EgorBo) [reply src/mono/mono/mini/mini-ops.h]: Renamed, in `mini-llvm.c` for `INTRINS_SSE_*` enum constants we don't use version suffix so I didn't change that

**Comment by tannergooding** [comment]
@EgorBo, is there a doc on how to add these as work happens so we can get both done at the same time in the future?

Is it worth doing it for ARM64 as well and would the process be similar?

**Comment by EgorBo** [comment]
> @EgorBo, is there a doc on how to add these as work happens so we can get both done at the same time in the future?

@tannergooding Not yet, the code needs some refactoring as there are too many repeatable patterns. The problem is the fact that sometime hw intrinsics are implemented as LLVM named intrinsics but in most cases we have to use native llvm-vector operators to implement them. But I'll try to compose a small doc. But first we need to add "Mono-LLVM CI lane" to catch such things (intrinsics are implemented only when LLVM back-end is available) 🙂 

> Is it worth doing it for ARM64 as well and would the process be similar?
We currently don't support them yet (there is a PR but it's WIP and only for scalar intrinsics so far)

**Pull request merged**